### PR TITLE
SIP279 SIPNET Restart MVP

### LINKED ISSUE #279: Implement MVP full-state restart support in SIPNET
We need restart support in SIPNET so that segmented runs can stop and resume without hidden-state drift.

Scope (MVP):

- Add restart input/output options to SIPNET runtime.
- Write a versioned restart checkpoint (`SIPNET_RESTART 1.0`, text format) containing all runtime state needed for deterministic resume.
- On restart, run normal setup, restore checkpointed state, and enforce strict restart compatibility checks.
- Preserve cumulative tracker continuity across segments.
- Keep output stitching in orchestration (no append/stitch logic in SIPNET).

Constraints:

- Resumed climate input must start at a timestamp (`year`, `day`, `time`) strictly after the checkpoint boundary timestamp.
- Do not allow replay/modification of already-processed events (deterministic cursor/history checks).
- Maintain backward behavior when restart options are not used.

Validation:

- Add hop-style tests comparing continuous vs segmented runs.
- Include fail-fast tests for restart boundary/climate ordering mismatch and invalid/corrupt restart inputs.
- Confirm non-restart behavior is unchanged.

Out of scope:

- Workflow/orchestration redesign.
- Mid-run parameter hot-swapping.
- Output merge/stitch features.

### PR DESCRIPTION
**Pull request by dlebauer**
## Summary

- **What**: Adds MVP full-state restart support for segmented SIPNET runs.
    - Introduces ASCII text restart file as restart schema v1.0 
    - Stops at end of .clim file and writes a restart file `RESTART_OUT`
    - Restarts by specifying `RESTART_IN` file
    - Writes one output file per climate segment (these outputs can be concatenated downstream)
- **Motivation**: 
    1. To support parameter changes to represent planting of a different PFT
    2. To support state data assimilation

## How was this change tested?

Ran:

- `make sipnet`
- `make testbuild`
- `make testrun`
- `make smoke`

Added restart test coverage:

- hop test: continuous vs segmented equivalence
- same-day multi-event boundary non-replay
- climate ordering 
- schema mismatch failure
- model/build mismatch failure
- truncated checkpoint failure
- no-restart unchanged behavior

### Key Changes

- Runtime restart module:
  - `src/sipnet/restart.c`
  - `src/sipnet/restart.h`
- Runtime wiring:
  - `src/sipnet/sipnet.c`
  - `src/common/context.c`
  - `src/sipnet/cli.c`
- Event determinism helpers:
  - `src/sipnet/events.c`
  - `src/sipnet/events.h`
- Tests:
  - `tests/sipnet/test_restart_infrastructure/testRestartMVP.c`
- Docs:
  - `docs/user-guide/running-sipnet.md`
  - `docs/developer-guide/restart-checkpoint.md`
  - `mkdocs.yml`

## Related Issues

- discussed in https://github.com/orgs/ccmmf/discussions/188 
- requirements in https://github.com/ccmmf/organization/issues/180
- requirements in #271 
- implementation spec, fixes #279

## Checklist

- [x] Related issues are listed above. [PRs without an approved, related issue may not get reviewed](docs/CONTRIBUTING.md#propose-and-receive-feedback).
- [x] PR title has the issue number in it ("[#<number>] \<concise description of proposed change>")
- [x] Tests added/updated for new features (if applicable)
- [x] Documentation updated (if applicable)
- [x] `docs/CHANGELOG.md` updated with noteworthy changes
- [x] Code formatted with `clang-format` (run `git clang-format` if needed)

## Notes

- Restart climate input must begin with a timestamp (`year`, `day`, `time`) that occurs after the last segment.
- Implementation and initial test scaffolding in this PR was generated with OpenAI Codex based on my specifications and review, but my C skills are limited. My specifications were based on issues linked above.

### CODE REVIEW

File: src/sipnet/sipnet.c [hunk 1]
```diff
@@ -228,21 +227,17 @@ void readClimData(const char *climFile) {
...
+      // trackers.gdd for restart continuity.
       thisGdd = tair * length;
       if (thisGdd < 0) {  // can't have negative growing degree days
         thisGdd = 0;
       }
-      gdd += thisGdd;
-      curr->gdd = gdd;
+      curr->gdd = thisGdd;
+    } else {
+      curr->gdd = 0.0;
```
Comment (Alomir) [inline src/sipnet/sipnet.c]: I'm tempted to just remove `ggd` from the climate struct... but this is ok, esp. given that the value is used in more than one spot.

File: docs/developer-guide/restart-checkpoint.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+## Climate and Event Boundaries
+
+Restart writes are only allowed when the last processed climate step is within one timestep of midnight.
+
+Resumed climate segments must begin on the day after the checkpoint boundary and within one timestep after midnight, where the window uses the first resumed climate row's timestep length.
+
+Event files must be segmented to the same time boundaries as climate segments.
+
+When `--gdd` is enabled, checkpoint resume restores cumulative GDD from `trackers.gdd`.
+`boundary.*` does not contain cumulative GDD.
```
Comment (Alomir) [inline docs/developer-guide/restart-checkpoint.md]: This is too down in the weeds of the implementation; there's no reason boundary would contain cumulative gdd without deep knowledge of past implementation.

File: docs/developer-guide/restart-checkpoint.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+## Struct Drift Guards
+
+Restart schema v1.0 includes compile-time and runtime drift guards so struct layout changes cannot silently pass:
+
+- Compile-time guards: `_Static_assert` checks in `src/sipnet/restart.c` for `Envi`, `Trackers`, `PhenologyTrackers`, and `EventTrackers`.
+- Runtime guards: `schema_layout.*` fields in each checkpoint are validated on load.
+- Test guardrails: `tests/sipnet/test_restart_infrastructure/testRestartMVP.c` verifies schema layout keys are present and rejects tampered values.
+
+## Schema Bump Checklist
```
Comment (Alomir) [inline docs/developer-guide/restart-checkpoint.md]: Would we like a section here on "what to do when altering a saved struct/adding data that needs to be saved"?

File: docs/user-guide/model-inputs.md [hunk 4]
```diff
@@ -211,17 +211,22 @@ Thus, command-line arguments override settings in the configuration file, and co
 | ------------ | --------- | ------------------------------------- |
 | `input-file` | sipnet.in | Name of input config file             |
 | `file-name`  | sipnet    | Prefix of climate and parameter files |
+| `events-file` | events   | Prefix for events input file (reads `<name>.in`) |
+| `restart-in` | unset     | Path to restart checkpoint to load    |
+| `restart-out` | unset    | Path to restart checkpoint to write   |
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply docs/user-guide/model-inputs.md]: Renamed the new events option/key to `events-prefix` / `EVENTS_PREFIX`, and it now drives both `<prefix>.in` and `<prefix>.out`. I left `file-name` alone here and opened #285 for the follow-up rename/alias discussion.

File: docs/user-guide/running-sipnet.md [hunk 5]
```diff
@@ -88,38 +94,42 @@ Keys are case-insensitive and can use hyphens or underscores (e.g., `EVENTS`, `e
...
-| `GROWTH_RESP` | 0 or 1 | Explicitly model growth respiration |
-| `LEAF_WATER` | 0 or 1 | Track separate leaf water pool |
-| `LITTER_POOL` | 0 or 1 | Enable separate litter pool |
-| `NITROGEN_CYCLE` | 0 or 1 | Enable nitrogen cycle modeling |
-| `SNOW` | 0 or 1 | Track snowpack |
-| `SOIL_PHENOL` | 0 or 1 | Use soil temperature for phenology |
-| `WATER_HRESP` | 0 or 1 | Allow soil moisture to affect respiration |
+| Key              | Value (1/0) | Description                               |
+| ---------------- | ----------- | ----------------------------------------- |
+| `EVENTS`         | 0 or 1      | Enable/disable event handling             |
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply docs/user-guide/running-sipnet.md]: Added `ANAEROBIC` to that table.

File: docs/CHANGELOG.md [hunk 6]
```diff
@@ -29,6 +29,7 @@ sections to include in release notes:
 ### Added
 
 - Build and release binaries for MacOS and Windows on release (in addition to existing Linux builds)
+- MVP restart checkpoints for segmented runs (`RESTART_IN` / `RESTART_OUT`) (#279)
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply docs/CHANGELOG.md]: Added a changelog line for the events prefix.

File: src/sipnet/cli.c [hunk 7]
```diff
@@ -74,6 +80,7 @@ void usage(char *progName) {
   printf("Options: (defaults are shown in parens at end)\n");
   printf("  -i, --input-file <input-file>      Name of input config file ('sipnet.in')\n");
   printf("  -f, --file-name  <name>            Prefix of climate and parameter files ('sipnet')\n");
+  printf("      --events-file <name>           Prefix of events input file ('events' => 'events.in')\n");
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/cli.c]: Updated this so the events prefix now controls both the input and output filenames.

File: src/sipnet/frontend.c [hunk 8]
```diff
@@ -186,7 +187,14 @@ int main(int argc, char *argv[]) {
   initModel(&modelParams, paramFile, climFile);
 
   if (ctx.events) {
-    initEvents(EVENT_IN_FILE, ctx.printHeader);
+    if (strlen(ctx.eventsFile) > FILENAME_MAXLEN - 4) {
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/frontend.c]: Implemented this in the frontend by constructing the event input/output filenames from `EVENTS_PREFIX` before initializing events.

File: src/sipnet/restart.c [hunk 9]
```diff
@@ -0,0 +1,1132 @@
...
+#include "version.h"
+
+#define RESTART_MAGIC "SIPNET_RESTART"
+#define RESTART_SCHEMA_VERSION "1.0"
+#define RESTART_FLOAT_EPSILON 1e-8
+
+#define RESTART_SCHEMA_LAYOUT_ENVI_SIZE 96
+#define RESTART_SCHEMA_LAYOUT_TRACKERS_SIZE 224
+#define RESTART_SCHEMA_LAYOUT_PHENOLOGY_TRACKERS_SIZE 12
+#define RESTART_SCHEMA_LAYOUT_EVENT_TRACKERS_SIZE 8
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/restart.c]: Updated these to use explicit `#define`s in `8 * (num elem)` style so the layout math is easier to read while debugging.

File: src/sipnet/restart.c [hunk 10]
```diff
@@ -0,0 +1,1132 @@
...
+
+_Static_assert(sizeof(Envi) == RESTART_SCHEMA_LAYOUT_ENVI_SIZE,
+               "Restart schema 1.0 drift: Envi changed; bump restart schema "
+               "version and update schema_layout.* checks");
+_Static_assert(sizeof(Trackers) == RESTART_SCHEMA_LAYOUT_TRACKERS_SIZE,
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/restart.c]: Yes, this was part of the issue. I folded in the PR 283 fix here by checking the serialized tracker payload shape rather than raw `Trackers` size.

File: src/sipnet/restart.c [hunk 11]
```diff
@@ -0,0 +1,1132 @@
...
+  int seenProcessedSteps = 0;
+  int seenSchemaLayout[4] = {0};
+
+  int seenFlags[10] = {0};
+  int seenBoundary[4] = {0};
+  int seenMeanMeta[5] = {0};
+  int seenEnvi[12] = {0};
+  int seenTrackers[28] = {0};
+  int seenPhenology[3] = {0};
+  int seenEventTrackers = 0;
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/restart.c]: Converted these to named schema-layout constants and kept the static checks tied to those `#define`s.

File: src/sipnet/restart.c [hunk 12]
```diff
@@ -0,0 +1,1132 @@
...
+#include "common/logging.h"
+#include "common/util.h"
+#include "events.h"
+#include "version.h"
+
+#define RESTART_MAGIC "SIPNET_RESTART"
+#define RESTART_SCHEMA_VERSION "1.0"
+#define RESTART_FLOAT_EPSILON 1e-8
+
+#define RESTART_SCHEMA_LAYOUT_ENVI_SIZE 96
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/restart.c]: Added a short near-code blurb covering what needs to change when the saved restart contract changes.

File: src/sipnet/restart.c [hunk 13]
```diff
@@ -0,0 +1,1132 @@
...
+  }
+}
+
+void restartResetRunState(void) {
+  processedStepCount = 0;
+  hasLastProcessedClimate = 0;
+}
+
+void restartNoteProcessedClimateStep(const ClimateNode *climateStep) {
+  copyClimateSignature(&lastProcessedClimate, climateStep);
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/restart.c]: Changed this to keep a pointer to the last processed climate step and copy into the serialized boundary only when writing the checkpoint.

File: src/sipnet/sipnet.c [hunk 14]
```diff
@@ -2076,7 +2076,9 @@ void runModelOutput(FILE *out, OutputItems *outputItems, int printHeader) {
     if (outputItems != NULL) {
       writeOutputItemValues(outputItems);
     }
-    restartNoteProcessedClimateStep(climate);
+    if (strlen(ctx.restartOut) > 0 || strlen(ctx.restartIn) > 0) {
+      restartNoteProcessedClimateStep(climate);
+    }
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply src/sipnet/sipnet.c]: Yes, narrowed this to `restartOut` only.

File: docs/developer-guide/restart-checkpoint.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+## Restart Schema v1.0 Overview
+
+Checkpoint format is ASCII text with one key/value per line:
+
+- header: `SIPNET_RESTART 1.0`
+- metadata: `model_version`, `build_info`, `checkpoint_utc_epoch`, `processed_steps`
+- schema layout guard metadata: `schema_layout.envi_size`, `schema_layout.trackers_size`, `schema_layout.phenology_trackers_size`, `schema_layout.event_trackers_size`
+  - `schema_layout.trackers_size` guards the serialized tracker payload shape for schema v1.0
+- mode flags: `flags.*`
+- boundary metadata: `boundary.year`, `boundary.day`, `boundary.time`, `boundary.length` (no forcing fields, no cumulative GDD)
```
[the remark this answers is not in the document]
Comment (dlebauer) [reply docs/developer-guide/restart-checkpoint.md]: Yes

File: docs/developer-guide/restart-checkpoint.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+On load, SIPNET enforces:
+
+- magic header match
+- schema version match
+- model numeric version match
+- `schema_layout.*` values exactly match the expected struct sizes for the running build
+- build info mismatch logs warning only
+- context flag compatibility
+- first-row climate timestamp strictly after checkpoint boundary (`year`, `day`, `time`)
+- resumed segment starts on the midnight-following day and within one timestep after midnight
```
[the remark this answers is not in the document]
Comment (infotroph) [reply docs/developer-guide/restart-checkpoint.md]: Part of the design of what, though? My point is that it seems like keeping the dump truly instantaneous without caring about the timesteps that preceded or followed it would _improve_ the design of the restart system. Are you saying the timestep complication is necessary because of the design of something else?
